course/auth.py

```diff
@@ -937,6 +935,28 @@ def adjust_layout(is_inst_id_locked):
         self.helper.add_input(
                 Submit("submit_user", _("Update")))
 
+        self.helper.add_input(
+                Button("signout", _("Sign out"), css_class="btn btn-danger",
+                       onclick=(
+                           "window.location.href=%s"
+                           % reverse("relate-logout"))))
+
+    @property
+    def changed_data(self):
+        data = super(UserForm, self).changed_data
+        may_not_change_field = []
+        if self.instance.name_verified:
+            may_not_change_field += ["first_name", "last_name"]
+
+        if self.is_inst_id_locked:
+            may_not_change_field.append("institutional_id")
+
+        changed_data = []
+        for name in self.Meta.fields:
+            if name in data and name not in may_not_change_field:
+                changed_data.append(name)
+        return changed_data
```

Comment:
Why not just fail validation if the data is changed (rather than pretend that the data didn't change)?

Reply:
That's a good point. What I did was ignore the change, and ignore saving of those items. Will update.

Reply:
I think we should raise those errors in `clean_` method.

Reply:
From the [doc](https://docs.djangoproject.com/en/dev/ref/forms/fields/#disabled)
> The disabled boolean argument, when set to True, disables a form field using the disabled HTML attribute so that it won’t be editable by users. Even if a user tampers with the field’s value submitted to the server, it will be ignored in favor of the value from the form’s initial data.

It seems we don't need to raise those errors. And, as you can see, those code were covered by tests, it seens the `changed_data` was generated before form data validation.

Reply:
From the [doc](https://docs.djangoproject.com/en/2.0/ref/forms/api/#django.forms.Form.changed_data):
>The `changed_data` attribute returns a list of the names of the fields whose values in the form’s bound data (usually `request.POST`) differ from what was provided in `initial`. It returns an empty list if no data differs.

The diff is between POST data, rather than cleaned_data. I overrided `changed_data` so that `has_changed` can be used to determine whether we need to do a DB save, and send what messages to user. If we want to diff `cleaned_data` with initial instance, we need to implement some new method.

Reply:
I just found out current implementation with old version of Django. Since 1.9 we an directly disable a field rather than disable its widget. Then the modification is rather simple.

Reply:
Updated.

Reply:
> I overrided changed_data so that has_changed can be used to determine whether we need to do a DB save

I still don't see how we're allowed to redefine the semantics of (and thus override) `changed_data` though--it's an interface defined by Django.

Reply:
You are right, my new idea is to remove that, by removing the checkbox field, and use a checkbox element instead.

Reply:
update.